File: Sources/Client/FrameTiming.cpp
#include "FrameTiming.h"
#include <cstring>

// Static member initialization
FrameTiming::TimePoint FrameTiming::s_frameStart;
FrameTiming::TimePoint FrameTiming::s_lastFrame;

double FrameTiming::s_deltaTime = 0.0;
double FrameTiming::s_accumulator = 0.0;
uint32_t FrameTiming::s_displayedThisSecond = 0;
uint32_t FrameTiming::s_fps = 0;
uint64_t FrameTiming::s_totalDisplayed = 0;

// Profiling static members
bool FrameTiming::s_profilingEnabled = false;
FrameTiming::TimePoint FrameTiming::s_stageStart[FrameTiming::STAGE_COUNT];
double FrameTiming::s_stageTimeMS[FrameTiming::STAGE_COUNT] = { 0 };
double FrameTiming::s_stageAccumMS[FrameTiming::STAGE_COUNT] = { 0 };
double FrameTiming::s_stageAvgMS[FrameTiming::STAGE_COUNT] = { 0 };
uint32_t FrameTiming::s_profileFrameCount = 0;

void FrameTiming::Initialize()
{
	s_lastFrame = Clock::now();
	s_deltaTime = 0.0;
	s_accumulator = 0.0;
	s_displayedThisSecond = 0;
	s_fps = 0;
	s_totalDisplayed = 0;

	// Initialize profiling
	s_profilingEnabled = false;
	s_profileFrameCount = 0;
	std::memset(s_stageTimeMS, 0, sizeof(s_stageTimeMS));
	std::memset(s_stageAccumMS, 0, sizeof(s_stageAccumMS));
	std::memset(s_stageAvgMS, 0, sizeof(s_stageAvgMS));
}

void FrameTiming::BeginFrame()
{
	s_frameStart = Clock::now();

	// Calculate delta time in seconds
	auto elapsed = std::chrono::duration<double>(s_frameStart - s_lastFrame);
	s_deltaTime = elapsed.count();

	// Clamp delta to prevent spiral of death (e.g., debugger pause)
	if (s_deltaTime > 0.25) s_deltaTime = 0.25;

	// Reset per-frame profiling times
	if (s_profilingEnabled)
	{
		std::memset(s_stageTimeMS, 0, sizeof(s_stageTimeMS));
		BeginProfile(ProfileStage::FrameTotal);
	}
}

void FrameTiming::EndFrame()
{
	// End total frame profiling
	if (s_profilingEnabled)
	{
		EndProfile(ProfileStage::FrameTotal);

		// Accumulate for averaging
		for (int i = 0; i < STAGE_COUNT; i++)
		{
			s_stageAccumMS[i] += s_stageTimeMS[i];
		}
		s_profileFrameCount++;
	}

	s_lastFrame = s_frameStart;
	s_accumulator += s_deltaTime;

	// Update FPS and profile averages every second
	if (s_accumulator >= 1.0)
	{
		s_fps = s_displayedThisSecond;
		s_displayedThisSecond = 0;
		s_accumulator -= 1.0;  // Preserve remainder for accuracy

		// Calculate profile averages
		if (s_profilingEnabled && s_profileFrameCount > 0)
		{
			for (int i = 0; i < STAGE_COUNT; i++)
			{
				s_stageAvgMS[i] = s_stageAccumMS[i] / s_profileFrameCount;
				s_stageAccumMS[i] = 0.0;
			}
			s_profileFrameCount = 0;
		}
	}
}

void FrameTiming::CountDisplayedFrame()
{
	// Called only when a frame is actually flipped/displayed
	s_totalDisplayed++;
	s_displayedThisSecond++;
}

double FrameTiming::GetDeltaTime()
{
	return s_deltaTime;
}

double FrameTiming::GetDeltaTimeMS()
{
	return s_deltaTime * 1000.0;
}

uint32_t FrameTiming::GetFPS()
{
	return s_fps;
}

uint64_t FrameTiming::GetFrameCount()
{
	return s_totalDisplayed;
}

// Profiling implementation
void FrameTiming::SetProfilingEnabled(bool enabled)
{
	if (enabled && !s_profilingEnabled)
	{
		// Reset when enabling
		s_profileFrameCount = 0;
		std::memset(s_stageTimeMS, 0, sizeof(s_stageTimeMS));
		std::memset(s_stageAccumMS, 0, sizeof(s_stageAccumMS));
		std::memset(s_stageAvgMS, 0, sizeof(s_stageAvgMS));
	}
	s_profilingEnabled = enabled;
}

bool FrameTiming::IsProfilingEnabled()
{
	return s_profilingEnabled;
}

void FrameTiming::BeginProfile(ProfileStage stage)
{
	if (!s_profilingEnabled) return;
	int idx = static_cast<int>(stage);
	if (idx >= 0 && idx < STAGE_COUNT)
	{
		s_stageStart[idx] = Clock::now();
	}
}

void FrameTiming::EndProfile(ProfileStage stage)
{
	if (!s_profilingEnabled) return;
	int idx = static_cast<int>(stage);
	if (idx >= 0 && idx < STAGE_COUNT)
	{
		auto now = Clock::now();
		auto elapsed = std::chrono::duration<double, std::milli>(now - s_stageStart[idx]);
		s_stageTimeMS[idx] += elapsed.count();  // += allows nested/multiple calls
	}
}

double FrameTiming::GetProfileTimeMS(ProfileStage stage)
{
	int idx = static_cast<int>(stage);
	if (idx >= 0 && idx < STAGE_COUNT)
		return s_stageTimeMS[idx];
	return 0.0;
}

double FrameTiming::GetProfileAvgTimeMS(ProfileStage stage)
{
	int idx = static_cast<int>(stage);
	if (idx >= 0 && idx < STAGE_COUNT)
		return s_stageAvgMS[idx];
	return 0.0;
}

const char* FrameTiming::GetStageName(ProfileStage stage)
{
	switch (stage)
	{
	case ProfileStage::Update:           return "Update";
	case ProfileStage::ClearBuffer:      return "ClearBuf";
	case ProfileStage::DrawBackground:   return "Background";
	case ProfileStage::DrawEffectLights: return "Lights";
	case ProfileStage::DrawObjects:      return "Objects";
	case ProfileStage::DrawEffects:      return "Effects";
	case ProfileStage::DrawWeather:      return "Weather";
	case ProfileStage::DrawChat:         return "Chat";
	case ProfileStage::DrawDialogs:      return "Dialogs";
	case ProfileStage::DrawMisc:         return "Misc";
	case ProfileStage::Flip:             return "Flip";
	case ProfileStage::FrameTotal:       return "Total";
	default:                             return "Unknown";
	}
}

File: engine/src/core/FrameTiming.ts
/**
 * FrameTiming — Frame timing, FPS tracking, and per-stage profiling.
 * Mirrors C++ FrameTiming from Sources/Client/FrameTiming.h/.cpp.
 *
 * Usage:
 *   FrameTiming.initialize();
 *   // In event loop:
 *   FrameTiming.beginFrame();
 *   game.renderFrame();       // calls beginProfile / endProfile internally
 *   FrameTiming.endFrame();
 */

// ── Profile stages (mirrors C++ ProfileStage enum) ────────────────

export enum ProfileStage {
	Update = 0,
	ClearBuffer,
	DrawBackground,
	DrawEffectLights,
	DrawObjects,
	DrawEffects,
	DrawWeather,
	DrawChat,
	DrawDialogs,
	DrawMisc,
	Flip,
	FrameTotal,
	COUNT,
}

const STAGE_NAMES: Record<ProfileStage, string> = {
	[ProfileStage.Update]:           'Update',
	[ProfileStage.ClearBuffer]:      'ClearBuffer',
	[ProfileStage.DrawBackground]:   'DrawBackground',
	[ProfileStage.DrawEffectLights]: 'DrawEffectLights',
	[ProfileStage.DrawObjects]:      'DrawObjects',
	[ProfileStage.DrawEffects]:      'DrawEffects',
	[ProfileStage.DrawWeather]:      'DrawWeather',
	[ProfileStage.DrawChat]:         'DrawChat',
	[ProfileStage.DrawDialogs]:      'DrawDialogs',
	[ProfileStage.DrawMisc]:         'DrawMisc',
	[ProfileStage.Flip]:             'Flip',
	[ProfileStage.FrameTotal]:       'FrameTotal',
	[ProfileStage.COUNT]:            '',
};

// ── FrameTiming singleton (all static) ────────────────────────────

const MAX_DELTA = 0.25; // Clamp to prevent spiral of death (seconds)

export class FrameTiming {
	// Timing
	private static s_lastFrameTime = 0;
	private static s_deltaTime = 0;       // seconds
	private static s_deltaTimeMS = 0;     // milliseconds

	// FPS
	private static s_fps = 0;
	private static s_displayedFrames = 0;
	private static s_fpsAccumulator = 0;
	private static s_fpsFrameCount = 0;
	private static s_totalFrameCount = 0;

	// Profiling
	private static s_profilingEnabled = false;
	private static s_profileStart = new Float64Array(ProfileStage.COUNT);
	private static s_profileTime = new Float64Array(ProfileStage.COUNT);   // current frame (ms)
	private static s_profileAccum = new Float64Array(ProfileStage.COUNT);  // accumulator for averaging
	private static s_profileAvg = new Float64Array(ProfileStage.COUNT);    // averaged (ms)
	private static s_profileFrames = 0;

	// ── Lifecycle ──────────────────────────────────────────────────

	static initialize(): void {
		FrameTiming.s_lastFrameTime = performance.now();
		FrameTiming.s_deltaTime = 0;
		FrameTiming.s_deltaTimeMS = 0;
		FrameTiming.s_fps = 0;
		FrameTiming.s_displayedFrames = 0;
		FrameTiming.s_fpsAccumulator = 0;
		FrameTiming.s_fpsFrameCount = 0;
		FrameTiming.s_totalFrameCount = 0;
		FrameTiming.s_profileFrames = 0;
	}

	/** Call at the start of each frame. Computes delta time. */
	static beginFrame(): void {
		const now = performance.now();
		let dt = (now - FrameTiming.s_lastFrameTime) / 1000; // to seconds
		if (dt > MAX_DELTA) dt = MAX_DELTA;
		FrameTiming.s_deltaTime = dt;
		FrameTiming.s_deltaTimeMS = dt * 1000;
		FrameTiming.s_lastFrameTime = now;

		// Reset per-frame profile times
		if (FrameTiming.s_profilingEnabled) {
			FrameTiming.s_profileTime.fill(0);
		}

		// Start FrameTotal profile
		FrameTiming.beginProfile(ProfileStage.FrameTotal);
	}

	/** Call at the end of each frame. Updates FPS counter. */
	static endFrame(): void {
		// End FrameTotal profile
		FrameTiming.endProfile(ProfileStage.FrameTotal);

		// Accumulate profile times for averaging
		if (FrameTiming.s_profilingEnabled) {
			for (let i = 0; i < ProfileStage.COUNT; i++) {
				FrameTiming.s_profileAccum[i] += FrameTiming.s_profileTime[i];
			}
			FrameTiming.s_profileFrames++;
		}

		// FPS: update every second
		FrameTiming.s_fpsAccumulator += FrameTiming.s_deltaTime;
		if (FrameTiming.s_fpsAccumulator >= 1.0) {
			FrameTiming.s_fps = FrameTiming.s_fpsFrameCount;
			FrameTiming.s_fpsFrameCount = 0;
			FrameTiming.s_fpsAccumulator -= 1.0;

			// Compute profile averages
			if (FrameTiming.s_profilingEnabled && FrameTiming.s_profileFrames > 0) {
				for (let i = 0; i < ProfileStage.COUNT; i++) {
					FrameTiming.s_profileAvg[i] =
						FrameTiming.s_profileAccum[i] / FrameTiming.s_profileFrames;
				}
				FrameTiming.s_profileAccum.fill(0);
				FrameTiming.s_profileFrames = 0;
			}
		}
	}

	/** Call only when a frame is actually displayed (after flip). */
	static countDisplayedFrame(): void {
		FrameTiming.s_totalFrameCount++;
		FrameTiming.s_fpsFrameCount++;
	}

	// ── Accessors ──────────────────────────────────────────────────

	/** Seconds since last frame (clamped to MAX_DELTA). */
	static getDeltaTime(): number {
		return FrameTiming.s_deltaTime;
	}

	/** Milliseconds since last frame. */
	static getDeltaTimeMS(): number {
		return FrameTiming.s_deltaTimeMS;
	}

	/** Displayed frames per second (updated once per second). */
	static getFPS(): number {
		return FrameTiming.s_fps;
	}

	/** Total displayed frames since initialization. */
	static getFrameCount(): number {
		return FrameTiming.s_totalFrameCount;
	}

	// ── Profiling ──────────────────────────────────────────────────

	static setProfilingEnabled(enabled: boolean): void {
		FrameTiming.s_profilingEnabled = enabled;
		if (enabled) {
			FrameTiming.s_profileTime.fill(0);
			FrameTiming.s_profileAccum.fill(0);
			FrameTiming.s_profileAvg.fill(0);
			FrameTiming.s_profileFrames = 0;
		}
	}

	static isProfilingEnabled(): boolean {
		return FrameTiming.s_profilingEnabled;
	}

	static beginProfile(stage: ProfileStage): void {
		if (!FrameTiming.s_profilingEnabled && stage !== ProfileStage.FrameTotal) return;
		FrameTiming.s_profileStart[stage] = performance.now();
	}

	static endProfile(stage: ProfileStage): void {
		if (!FrameTiming.s_profilingEnabled && stage !== ProfileStage.FrameTotal) return;
		FrameTiming.s_profileTime[stage] += performance.now() - FrameTiming.s_profileStart[stage];
	}

	/** Current frame time for a stage (ms). */
	static getProfileTimeMS(stage: ProfileStage): number {
		return FrameTiming.s_profileTime[stage];
	}

	/** Averaged time for a stage over ~1 second (ms). */
	static getProfileAvgTimeMS(stage: ProfileStage): number {
		return FrameTiming.s_profileAvg[stage];
	}

	static getStageName(stage: ProfileStage): string {
		return STAGE_NAMES[stage] ?? '';
	}
}
